### ISSUE TITLE
Grafana: provision data in YAML

### ISSUE THREAD

**Issue by KFearsoff** [NixOS/nixpkgs#191464]
## Issue description

Grafana has a feature called [provisioning](https://grafana.com/docs/grafana/latest/administration/provisioning/), which lets an administrator specify Grafana data declaratively. It is immensely useful, however, we implement it by hand, writing a ton of options by hand (pretty much all of `services.grafana.provision.dashboards.*`, for example). This has a number of drawbacks.

### Drawbacks

1. We end up duplicating effort, having to write NixOS module options for every option in original YAML, while we could just use the original YAML from the get-go.
2. We have to continiously track the upstream to add new options. Failure to do so leads to subpar experience on NixOS due to missing options. For example, we don't implement a `disableDeletion` option for dashboard provisioning as of now, which is genuinely useful.
3. Writing modules where we could just use YAML leads to inflated line counts and option counts, which just overcomplicates things needlessly.
4. It's bad for system administrators who want to migrate to NixOS: instead of directly importing YAML configs, they'd have to rewrite a lot of stuff in a new DSL, which is time-consuming.

### Discussion

There are some arguments for keeping things as they are, though. Using modules lets us do Nix-side validation and Nix-side documentation. Also, it allows us to implement custom logic.
Ultimately, I think those advantages are not worth it, as we'd have to essentially duplicate upstream validation and upstream documentation for no good reason. Also, it's not like using YAML doesn't let us implement custom logic: we could use `builtins.fromJSON` to parse the YAML to get a valid Nix struct, which we then would do logic on. It is by no means a perfect solution, as that would take more compute and lead to writing complicated functions to implement the same logic, but in a lot of cases (such as Grafana), we don't even use any complicated logic to begin with.
Having custom logic isn't useless, though, so I could easily see giving users a choice.

### Details

We could implement it like this:
```nix
options.services.grafana.provision.dashboards = mkOption {
  type = with types; oneOf [ path (listOf attrsOf anything) ];
  description = mkDoc "Path to YAML or Nix implementation of YAML according to Grafana docs";
};

config = {
  let 
    provisionedDashboards = if (typeOf cfg.provision.dashboards == "path) then builtins.readFile cfg.provision.dashboards else builtins.fromJson cfg.provision.dashboards;
  in
    ...
};
```
This implementation allows us to completely sidestep the problem of invalidating all of the configurations that currently use the module system we have in place (mostly), while also allowing us to use YAML configuration files instead. 

## Extra notes

This is not directly addressed by #144575, since Grafana module doesn't have an `extraConfig`. It is a similar problem, though.
Also, this might warrant an RFC, since we have a number of modules that are in the same situation. YAML is pretty much a standard configuration format nowadays, it would be better to support is as a first-class citizen to ease the transition to NixOS.

**Comment by Ma27** [comment]
> This is not directly addressed by https://github.com/NixOS/nixpkgs/issues/144575, since Grafana module doesn't have an extraConfig

I disagree here: yes, one of the original motivations for RFC42 was to get rid of extraConfig, but IMHO this is a sufficient, but not necessary property to be a case of RFC42. In the end it's a (pretty reasonable) style of writing modules.

In fact, you're listing a few issues that RFC42 also addresses, such as having to track upstream options.

Now to the motivation of the issue itself: first of all a disclaimer, I don't know from memory how to configure Grafana without the NixOS module, so some minor details might not be completely correct, but:

* we could implement a `services.grafana.settings` for everything that belongs in e.g. `grafana.ini`
* we could add `services.grafana.provisioning.settings` that the structure of provisioning settings for dashboards/datasoruces.
* for both trees we should use a `freeformType` to make sure that we don't have to track every single config option. However I think we should keep the existing options since we already have docs adn proper type checking for that (and it would be a waste to throw that away)
* as you already mentioned, we could also add `services.grafana.provisioning.path` which points to a directory containing yaml files.
* I guess we could even allow both `path` and `settings` and create a store-path that contains a JSON-file with the Nix attr-set from `services.grafana.provisioning.settings` and YAML files from `services.grafana.provisioning.path`.
* If any of these options in `services.grafana.provisioning` is set, `services.grafana.settings.paths.provisioning` should be automatically set (which is already the case, but with an environment variable `GF_PATHS_PROVISIONING`).

Additionally there are a few things to look out for:
* I really dislike it when we break APIs just because we can and not because of an actual need (such as breaking changes). I operate a big fleet of NixOS machines for a living and there are a lot of challenges when updating NixOS for us, so these kinds of API changes are just an (IMHO unnecessary burden). This means that we should aggressively use `mkRenamedOptionModule` to make sure that all top-level settings are properly mapped to `settings` (i.e. `services.grafana.port` is mapped to `services.grafana.settings.server.http_port`). I'm fine with removing this layer after a few NixOS releases, but I think throwing the entire API away just now is way too harsh.
* The same thing applies to `services.grafana.provision.{datasources,dashbaords,notifiers}`, of course.
* We have to check for potential conflicts between `extraOptions` (i.e. custom settings that are passed as environment vars) and our new settings facility. This means that `extraOptions` should take precedence over settings (or a big warning is shown in the release notes and on evaluation that `extraOptions` are potentially overridden now).

This is just a braindump of my current thoughts, not a completely thought-out suggestion. Does this make sense to you?

**Comment by KFearsoff** [comment]
Yes, this makes a lot of sense, thank you for your input!
> However I think we should keep the existing options since we already have docs adn proper type checking for that (and it would be a waste to throw that away)

I think this is a two-fold problem. I currently haven't yet looked at Grafana settings, only on provisioning settings. In case of provisioning settings, we don't need to specify any options explicitly, because the names map 1:1 to the Grafana settings names. I don't know if that's the case for the Grafana settings.
Anyway, I had a discussion about that in NixOS Matrix server, and I was suggested to only keep the most used options around, and everything else should be a part of a `freeformType`. I think that makes sense: from what I understand, RFC42 suggests to remove the options that aren't often used. I am planning on only leaving the most used options, given that doesn't break backwards compatibility.
> I guess we could even allow both `path` and `settings`

I am not sure about that. I think the `services.grafana.provisioning.path` and `services.grafana.provisioning.settings` API makes sense. I think it will make for a much harder transition, though: if we can afford to switch to a `services.grafana.provisioning` that would take in both `path` and `settings` without breaking the API and them emitting a warning to encourage switching to `services.grafana.provisioning.settings` would be best.


As for other things, yes I agree. I will do my best to preserve backwards compatibility. It would also be very nice to have you look at the PR I'll come up with once it's up.

I'll leave the issue hanging for future discussion regarding Grafana specifically.

**Comment by Ma27** [comment]
> we don't need to specify any options explicitly, because the names map 1:1 to the Grafana settings names

I'd still keep those since we get type checking and documentation for free.

> from what I understand, RFC42 suggests to remove the options that aren't often used

I think that this should be considered on a case-by-case basis: unless an option is completely obscure I'd say it doesn't hurt considering that we **already have them written down**. If you'd implement the grafana module just now, it'd be perfectly understandable if you would've left out most of them.

> if we can afford to switch to a services.grafana.provisioning that would take in both path and settings without breaking the API and them emitting a warning to encourage switching to services.grafana.provisioning.settings would be best

I guess I wasn't clear enough here: that is exactly what I want as well ;-)
This can probably be implemented by mapping everything from `services.grafana.provision` (i.e. the existing options) to `services.grafana.provisioning.settings` (perhaps a single `mkRenamedOptionModule` is sufficient here).
Btw I just realized that I talked about `services.grafana.provisioning` even though `services.grafana.provision` is currently used. 
Not sure if we want to put `path`/`settings` into `provision` or use `provisioning` (that would probably be kinda confusing, OTOH consistent with upstream's naming and probably easier to implement), so I'd leave that open for discussion.

> It would also be very nice to have you look at the PR I'll come up with once it's up.

Yes, of course. In the end I'm one of the maintainers of `grafana` in nixpkgs %)

As soon as this has reached a usable state, I'd probably beta-test it against my personal setup where I'm also using a few advanced features of the grafana module.

**Comment by KFearsoff** [comment]
I've put out a draft PR at #191768, currently only the dashboard provisioning is done. I went with the "`services.grafana.provision.dashboards`still works, but `services.grafana.provision.dashboards.settings` or `services.grafana.provision.dashboards.path` is preferred" approach. It should be fully backwards compatible.

I think the approach itself is scalable. The NixOS Grafana module is pretty well-written, and the provision types are very similar to each other, so if the dashboard example works, then every other provision type should work as well. One downside is that the module became a lot less clean, but I don't think it's avoidable while preserving backwards compatibility, or I don't have enough skill to tell how to refactor it more elegantly.

Do you mind taking a look?

**Comment by KFearsoff** [comment]
@Ma27 The #191768  PR is out of the drafting phase, since all essential features I wanted to add are there. I would be very interested in what results you could arrive at if you tried the updated module, since I am very sure there are a lot of edge cases that you could probably spot very soon.

**Comment by KFearsoff** [comment]
#191768 is merged, so the issue is closed.
